pydra/engine/submitter.py

```diff
@@ -149,7 +150,9 @@ def __exit__(self, type, value, traceback):
             return futures
 
     def close(self):
-        self.loop.close()
+        # jupyter is also using cf, so can't close the loop
+        if not "Jupyter" in os.environ:
```

Comment:
this seems something that Jupyter should handle, not us. i don't want pydra to be dependent on whether it's used in a specific environment.

Reply:
The short answer is - it also surprised me, but Jupyter doesn't handle... Note, that it's not only this, I still had to import `nest_asyncio` to the notebook!

I agree that we don't want to change pydra, the if statement seemed to me the only "acceptable" option, but I will search more.

Reply:
yesterday I was thinking about forgetting about jupyter, but I'm guessing it's not an option that will be approved ;-)

Reply:
Do we have config?

```suggestion
        if not config.getboolean("execution", "nested_asyncio"):
```

Then, Jupyter notebooks could use a modified config file, or users could add at import: 

```Python
import pydra
pydra.config.set("execution", "nested_asyncio", True)
```

Reply:
but the problem is that we also want pydra to change the behavior, i.e. do not close the loop. At least that was the solution we found yesterday

Reply:
one possibility is for pydra to check if there is a loop and reuse it rather than creating a new one? and, in such an event, not close it.

Reply:
we were thinking about it yesterday, but we were afraid that this might happen also in pydra Workflow itself, and has nothing to do with jupyter.

Just to be clear all tests pass without closing the loop at all, it's "just" a best practice, and I'm not sure when this might be a problem.

@mgxd can comment as well

Reply:
a) a loop should be opened only when the program is not already in a loop
b) a loop should be closed only if the loop has been opened by the program

Reply:
we can easily detect if the available loop is currently running, and avoid closing it, but we'll still need to import / activate `nest_asyncio` to allow running tasks through jupyter

Reply:
@satra - looks like this is not issue create by jupyter, but newer version of tornado, and it looks like this is still not fixed. It works fine (at least for my simple example, will be testing more) if older version of tornado is used `tornado==4.5.3`

Reply:
didn't read all posts, but you can find more here: https://github.com/jupyter/notebook/issues/3397